From f04cff6a605661cfe42c03a290ef092a465d3cbb Mon Sep 17 00:00:00 2001
From: Chris Yeh <chris_yeh@phoenix.com>
Date: Thu, 2 Dec 2021 20:55:10 +0800
Subject: [PATCH] Sync intel peci patches

- 0502-mfd-peci-Add-SPR-generation-info.patch
- 0503-hwmon-peci-platformpower-implementation.patch
- 0506-peci-cputemp-Add-HBM-temperature-sensor-support.patch
---
 Documentation/hwmon/index.rst                 |   1 +
 Documentation/hwmon/peci-platformpower.rst    |  61 ++
 .../dts/aspeed-bmc-phoenix-brightoncity.dts   | 527 ++++++++++
 arch/arm/boot/dts/aspeed-g5.dtsi              |  26 +-
 drivers/hwmon/Kconfig                         |  15 +
 drivers/hwmon/Makefile                        |   1 +
 drivers/hwmon/peci-cputemp.c                  | 106 +-
 drivers/hwmon/peci-dimmtemp.c                 |  66 ++
 drivers/hwmon/peci-platformpower.c            | 910 ++++++++++++++++++
 drivers/mfd/intel-peci-client.c               |   7 +
 include/linux/mfd/intel-peci-client.h         |   7 +-
 11 files changed, 1705 insertions(+), 22 deletions(-)
 create mode 100644 Documentation/hwmon/peci-platformpower.rst
 create mode 100644 arch/arm/boot/dts/aspeed-bmc-phoenix-brightoncity.dts
 create mode 100644 drivers/hwmon/peci-platformpower.c

diff --git a/Documentation/hwmon/index.rst b/Documentation/hwmon/index.rst
index b1379e47ba6f..8676c610dd0d 100644
--- a/Documentation/hwmon/index.rst
+++ b/Documentation/hwmon/index.rst
@@ -148,6 +148,7 @@ Hardware Monitoring Kernel Drivers
    peci-dimmtemp
    peci-cpupower
    peci-dimmpower
+   peci-platformpower
    pmbus
    powr1220
    pxe1610
diff --git a/Documentation/hwmon/peci-platformpower.rst b/Documentation/hwmon/peci-platformpower.rst
new file mode 100644
index 000000000000..d814dda91dcc
--- /dev/null
+++ b/Documentation/hwmon/peci-platformpower.rst
@@ -0,0 +1,61 @@
+.. SPDX-License-Identifier: GPL-2.0
+
+Kernel driver peci-platformpower
+==========================
+
+:Copyright: |copy| 2020 Intel Corporation
+
+Supported chips:
+	One of Intel server CPUs listed below which is connected to a PECI bus.
+		* Intel Xeon E5/E7 v3 server processors
+			Intel Xeon E5-14xx v3 family
+			Intel Xeon E5-24xx v3 family
+			Intel Xeon E5-16xx v3 family
+			Intel Xeon E5-26xx v3 family
+			Intel Xeon E5-46xx v3 family
+			Intel Xeon E7-48xx v3 family
+			Intel Xeon E7-88xx v3 family
+		* Intel Xeon E5/E7 v4 server processors
+			Intel Xeon E5-16xx v4 family
+			Intel Xeon E5-26xx v4 family
+			Intel Xeon E5-46xx v4 family
+			Intel Xeon E7-48xx v4 family
+			Intel Xeon E7-88xx v4 family
+		* Intel Xeon Scalable server processors
+			Intel Xeon D family
+			Intel Xeon Bronze family
+			Intel Xeon Silver family
+			Intel Xeon Gold family
+			Intel Xeon Platinum family
+
+	Addresses scanned: PECI client address 0x30
+	Datasheet: Available from http://www.intel.com/design/literature.htm
+
+Author:
+	Zbigniew Lukwinski <zbigniew.lukwinski@linux.intel.com>
+
+Description
+-----------
+
+This driver implements a generic PECI power monitoring feature which provides
+average power and energy consumption readings of the platform based on energy
+counter.
+Power values are average power since last measure given in milli Watt and
+will be measurable only when the target CPU is powered on.
+Energy values are cumulative energy consumption in micro Joules.
+Driver provides current platform power limit, maximal and minimal power
+setting as well.
+All needed processor registers are accessible using the PECI Client Command
+Suite via the processor PECI client.
+
+``sysfs`` interface
+-------------------
+======================= =======================================================
+power1_label		Provides string "platform power".
+power1_average		Provides average power since last read in milli Watt.
+power1_cap		Provides current platform power limit 1 (PPL1).
+power1_cap_max		Provides maximal platform power setting.
+power1_cap_min		Provides minimal platform power setting.
+energy1_label		Provides string "platform energy".
+energy1_input		Provides energy consumption in micro Joules
+======================= =======================================================
diff --git a/arch/arm/boot/dts/aspeed-bmc-phoenix-brightoncity.dts b/arch/arm/boot/dts/aspeed-bmc-phoenix-brightoncity.dts
new file mode 100644
index 000000000000..8493e6cdf0ab
--- /dev/null
+++ b/arch/arm/boot/dts/aspeed-bmc-phoenix-brightoncity.dts
@@ -0,0 +1,527 @@
+/dts-v1/;
+
+#include "aspeed-g5.dtsi"
+#include <dt-bindings/gpio/aspeed-gpio.h>
+#include <dt-bindings/i2c/i2c.h>
+
+/ {
+	model = "Phoenix AST2500 BMC";
+	compatible = "phoenix,ast2500-bmc", "aspeed,ast2500";
+
+	aliases {
+		serial4 = &uart5;
+	};
+
+	chosen {
+		stdout-path = &uart5;
+		bootargs = "console=ttyS4,115200 earlyprintk";
+	};
+
+	memory@80000000 {
+		reg = <0x80000000 0x20000000>;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		vga_memory: framebuffer@7f000000 {
+			no-map;
+			reg = <0x7f000000 0x01000000>;
+		};
+
+		gfx_memory: framebuffer {
+			size = <0x01000000>;
+			alignment = <0x01000000>;
+			compatible = "shared-dma-pool";
+			reusable;
+		};
+
+		video_engine_memory: jpegbuffer {
+			size = <0x02000000>;	/* 32M */
+			alignment = <0x01000000>;
+			compatible = "shared-dma-pool";
+			reusable;
+		};
+
+		fw_update: fw_update@9efe0000{
+			no-map;
+			reusable;
+			reg = <0x9efe0000 0x10000>;     /* 64k window */
+		};
+
+		ramoops@9eff0000{
+			compatible = "ramoops";
+			reg = <0x9eff0000 0x10000>;
+			record-size = <0x2000>;
+			console-size = <0x2000>;
+		};
+	};
+
+	vga-shared-memory {
+		compatible = "aspeed,ast2500-vga-sharedmem";
+		reg = <0x9ff00000 0x100000>;
+	};
+
+	iio-hwmon {
+		compatible = "iio-hwmon";
+		io-channels = <&adc 0>, <&adc 1>, <&adc 2>, <&adc 3>,
+			<&adc 4>, <&adc 5>, <&adc 6>, <&adc 7>,
+			<&adc 8>, <&adc 9>, <&adc 10>, <&adc 11>,
+			<&adc 12>, <&adc 13>, <&adc 14>, <&adc 15>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		identify {
+			default-state = "on";
+			gpios = <&gpio ASPEED_GPIO(S, 6) GPIO_ACTIVE_LOW>;
+		};
+
+		status_amber {
+			default-state = "off";
+			gpios = <&gpio ASPEED_GPIO(S, 5) GPIO_ACTIVE_LOW>;
+		};
+
+		status_green {
+			default-state = "keep";
+			gpios = <&gpio ASPEED_GPIO(S, 4) GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	beeper {
+		compatible = "pwm-beeper";
+		pwms = <&timer 6 1000000 0>;
+	};
+};
+
+&p2a {
+	status = "okay";
+	memory-region = <&fw_update>;
+};
+
+&fmc {
+	status = "okay";
+	flash@0 {
+		status = "okay";
+		m25p,fast-read;
+		label = "bmc";
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			u-boot@0 {
+				reg = <0x0 0xC0000>;
+				label = "u-boot";
+			};
+
+			u-boot-env@C0000 {
+				reg = <0xC0000 0x40000>;
+				label = "u-boot-env";
+			};
+
+			kernel@C0000 {
+				reg = <0x100000 0x700000>;
+				label = "kernel";
+			};
+
+			rofs@c0000 {
+				reg = <0x800000 0x2800000>;
+				label = "rofs";
+			};
+
+			rwfs@3000000 {
+				reg = <0x3000000 0x1000000>;
+				label = "rwfs";
+			};
+		};
+	};
+};
+
+&espi {
+	status = "okay";
+};
+
+&jtag {
+	status = "okay";
+};
+
+&peci0 {
+	status = "okay";
+	clock-frequency =  <2000000>;
+
+	peci-client@30 {
+		compatible = "intel,peci-client";
+		reg = <0x30>;
+	};
+
+};
+
+&syscon {
+	uart-clock-high-speed;
+	status = "okay";
+
+	misc_control {
+		compatible = "aspeed,bmc-misc";
+		uart_port_debug {
+			offset = <0x2c>;
+			bit-mask = <0x1>;
+			bit-shift = <10>;
+			read-only;
+		};
+		boot-2nd-flash {
+			offset = <0x70>;
+			bit-mask = <0x1>;
+			bit-shift = <17>;
+			read-only;
+		};
+		chip_id {
+			offset = <0x150>;
+			bit-mask = <0x0fffffff 0xffffffff>;
+			bit-shift = <0>;
+			read-only;
+			reg-width = <64>;
+			hash-data = "d44f9b804976fa23c2e25d62f16154d26520a7e24c5555095fd1b55c027804f1570dcd16189739c640cd7d9a6ce14944a2c4eaf1dc429eed6940e8a83498a474";
+		};
+	};
+};
+
+&adc {
+	status = "okay";
+};
+
+&gpio {
+	status = "okay";
+	/* Enable GPIOE0 and GPIOE2 pass-through by default */
+	pinctrl-names = "pass-through";
+	pinctrl-0 = <&pinctrl_gpie0_default
+			&pinctrl_gpie2_default>;
+	gpio-line-names =
+	/*A0-A7*/	"","","","","","","","",
+	/*B0-B7*/	"","","TP_BMC_GPIOB2","TP_BMC_GPIOB3","","","PU_BMC_GPIOB6","TP_BMC_GPIOB7",
+	/*C0-C7*/	"","","","","","","","",
+	/*D0-D7*/	"","","","","","","","",
+	/*E0-E7*/	"RESET_BUTTON","RESET_OUT","POWER_BUTTON","POWER_OUT","","DEBUG_EN_N","","",
+	/*F0-F7*/	"NMI_OUT","","","","CPU_ERR0","CPU_ERR1","PLTRST_N","PRDY_N",
+	/*G0-G7*/	"CPU_ERR2","CPU_CATERR","PCH_BMC_THERMTRIP","LCP_ENTER_BUTTON","LCP_LEFT_BUTTON","FM_BMC_BOARD_SKU_ID5_N","","",
+	/*H0-H7*/	"","","","FM_NODE_ID_1","FM_NODE_ID_2","FM_NODE_ID_3","FM_NODE_ID_4","FM_240VA_STATUS",
+	/*I0-I7*/	"","","","","","","","",
+	/*J0-J7*/	"","","","","","","","",
+	/*K0-K7*/	"","","","","","","","",
+	/*L0-L7*/	"","","","","","","","",
+	/*M0-M7*/	"","","","","","","","",
+	/*N0-N7*/	"","","","","","","","",
+	/*O0-O7*/	"","","","","","","","",
+	/*P0-P7*/	"","","","","","","","",
+	/*Q0-Q7*/	"","","","","","","","PWR_DEBUG_N",
+	/*R0-R7*/	"","XDP_PRST_N","","","","","","CHASSIS_INTRUSION",
+	/*S0-S7*/	"REMOTE_DEBUG_ENABLE","SYSPWROK","RSMRST_N","","","","","",
+	/*T0-T7*/	"","","","","","","","",
+	/*U0-U7*/	"","","","","","","","",
+	/*V0-V7*/	"","","","","","","","",
+	/*W0-W7*/	"","","","","","","","",
+	/*X0-X7*/	"","","","","","","","",
+	/*Y0-Y7*/	"SIO_S3","SIO_S5","","SIO_ONCONTROL","","","","",
+	/*Z0-Z7*/	"","SIO_POWER_GOOD","","","","","","",
+	/*AA0-AA7*/	"P3VBAT_BRIDGE_EN","","","","PREQ_N","TCK_MUX_SEL","SMI","POST_COMPLETE",
+	/*AB0-AB7*/	"","NMI_BUTTON","ID_BUTTON","PS_PWROK","","","","",
+	/*AC0-AC7*/	"","","","","","","","";
+};
+
+&sgpio {
+	ngpios = <80>;
+	status = "okay";
+	/* Enable GPIOE0 and GPIOE2 pass-through by default */
+	pinctrl-names = "pass-through";
+	pinctrl-0 = <&pinctrl_gpie0_default
+			&pinctrl_gpie2_default>;
+	gpio-line-names =
+	/* SGPIO output lines */
+	/*OA0-OA7*/	"","","","","","","","",
+	/*OB0-OB7*/	"LED_CPU1_CH1_DIMM1_FAULT","LED_CPU1_CH1_DIMM2_FAULT","LED_CPU1_CH2_DIMM1_FAULT","LED_CPU1_CH2_DIMM2_FAULT","LED_CPU1_CH3_DIMM1_FAULT","LED_CPU1_CH3_DIMM2_FAULT","LED_CPU1_CH4_DIMM1_FAULT","LED_CPU1_CH4_DIMM2_FAULT",
+	/*OC0-OC7*/	"LED_CPU1_CH5_DIMM1_FAULT","LED_CPU1_CH5_DIMM2_FAULT","LED_CPU1_CH6_DIMM1_FAULT","LED_CPU1_CH6_DIMM2_FAULT","LED_FAN1_FAULT","LED_FAN2_FAULT","LED_FAN3_FAULT","LED_FAN4_FAULT",
+	/*OD0-OD7*/	"LED_FAN5_FAULT","LED_FAN6_FAULT","LED_FAN7_FAULT","LED_FAN8_FAULT","LED_CPU2_CH1_DIMM1_FAULT","LED_CPU1_CH1_DIMM2_FAULT","LED_CPU2_CH2_DIMM1_FAULT","LED_CPU2_CH2_DIMM2_FAULT",
+	/*OE0-OE7*/	"LED_CPU2_CH3_DIMM1_FAULT","LED_CPU2_CH3_DIMM2_FAULT","LED_CPU2_CH4_DIMM1_FAULT","LED_CPU2_CH4_DIMM2_FAULT","LED_CPU2_CH5_DIMM1_FAULT","LED_CPU2_CH5_DIMM2_FAULT","LED_CPU2_CH6_DIMM1_FAULT","LED_CPU2_CH6_DIMM2_FAULT",
+	/*OF0-OF7*/	"LED_CPU3_CH1_DIMM1_FAULT","LED_CPU3_CH1_DIMM2_FAULT","LED_CPU3_CH2_DIMM1_FAULT","LED_CPU3_CH2_DIMM2_FAULT","LED_CPU3_CH3_DIMM1_FAULT","LED_CPU3_CH3_DIMM2_FAULT","LED_CPU3_CH4_DIMM1_FAULT","LED_CPU3_CH4_DIMM2_FAULT",
+	/*OG0-OG7*/	"LED_CPU3_CH5_DIMM1_FAULT","LED_CPU3_CH5_DIMM2_FAULT","LED_CPU3_CH6_DIMM1_FAULT","LED_CPU3_CH6_DIMM2_FAULT","LED_CPU4_CH1_DIMM1_FAULT","LED_CPU4_CH1_DIMM2_FAULT","LED_CPU4_CH2_DIMM1_FAULT","LED_CPU4_CH2_DIMM2_FAULT",
+	/*OH0-OH7*/	"LED_CPU4_CH3_DIMM1_FAULT","LED_CPU4_CH3_DIMM2_FAULT","LED_CPU4_CH4_DIMM1_FAULT","LED_CPU4_CH4_DIMM2_FAULT","LED_CPU4_CH5_DIMM1_FAULT","LED_CPU4_CH5_DIMM2_FAULT","LED_CPU4_CH6_DIMM1_FAULT","LED_CPU4_CH6_DIMM2_FAULT",
+	/*OI0-OI7*/	"","","","","","","","",
+	/*OJ0-OJ7*/	"","","","","","","","",
+	/*DUMMY*/	"","","","","","","","",
+	/*DUMMY*/	"","","","","","","","",
+
+	/* SGPIO input lines */
+	/*IA0-IA7*/	"CPU1_PRESENCE","CPU1_THERMTRIP","CPU1_VRHOT","CPU1_FIVR_FAULT","CPU1_MEM_ABCD_VRHOT","CPU1_MEM_EFGH_VRHOT","","",
+	/*IB0-IB7*/	"CPU1_MISMATCH","CPU1_MEM_THERM_EVENT","CPU2_PRESENCE","CPU2_THERMTRIP","CPU2_VRHOT","CPU2_FIVR_FAULT","CPU2_MEM_ABCD_VRHOT","CPU2_MEM_EFGH_VRHOT",
+	/*IC0-IC7*/	"","","CPU2_MISMATCH","CPU2_MEM_THERM_EVENT","","","","",
+	/*ID0-ID7*/	"","","","","","","","",
+	/*IE0-IE7*/	"","","","","","","","",
+	/*IF0-IF7*/	"SGPIO_PLD_MINOR_REV_BIT0","SGPIO_PLD_MINOR_REV_BIT1","SGPIO_PLD_MINOR_REV_BIT2","SGPIO_PLD_MINOR_REV_BIT3","SGPIO_PLD_MAJOR_REV_BIT0","SGPIO_PLD_MAJOR_REV_BIT1","SGPIO_PLD_MAJOR_REV_BIT2","SGPIO_PLD_MAJOR_REV_BIT3",
+	/*IG0-IG7*/	"MAIN_PLD_MINOR_REV_BIT0","MAIN_PLD_MINOR_REV_BIT1","MAIN_PLD_MINOR_REV_BIT2","MAIN_PLD_MINOR_REV_BIT3","MAIN_PLD_MAJOR_REV_BIT0","MAIN_PLD_MAJOR_REV_BIT1","MAIN_PLD_MAJOR_REV_BIT2","MAIN_PLD_MAJOR_REV_BIT3",
+	/*IH0-IH7*/	"","","","","","","","",
+	/*II0-II7*/	"","","","","","","","",
+	/*IJ0-IJ7*/	"","","","","","","","";
+};
+
+&kcs3 {
+	kcs_addr = <0xCA2>;
+	status = "okay";
+};
+
+&kcs4 {
+	kcs_addr = <0xCA4>;
+	status = "okay";
+};
+
+&sio_regs {
+	status = "okay";
+	sio_status {
+		offset = <0x8C>;
+		bit-mask = <0x1F>;
+		bit-shift = <4>;
+	};
+};
+
+&lpc_sio {
+	status = "okay";
+};
+
+&lpc_snoop {
+	snoop-ports = <0x80>;
+	status = "okay";
+};
+
+&mbox {
+	status = "okay";
+};
+
+&uart_routing {
+	status = "okay";
+};
+
+/**
+ * SAFS through SPI1 is available only on Wilson Point.
+ * These pins are used as fan presence checking gpios in WFP
+ * so commenting it out for now.
+ * &spi1 {
+ *	status = "okay";
+ *
+ *	flash@0 {
+ *		m25p,fast-read;
+ *		status = "okay";
+ *	};
+ *};
+ */
+
+&uart1 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_txd1_default
+			&pinctrl_rxd1_default
+			&pinctrl_nrts1_default
+			&pinctrl_ndtr1_default
+			&pinctrl_ndsr1_default
+			&pinctrl_ncts1_default
+			&pinctrl_ndcd1_default
+			&pinctrl_nri1_default>;
+};
+
+&uart2 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_txd2_default
+			&pinctrl_rxd2_default
+			&pinctrl_nrts2_default
+			&pinctrl_ndtr2_default
+			&pinctrl_ndsr2_default
+			&pinctrl_ncts2_default
+			&pinctrl_ndcd2_default
+			&pinctrl_nri2_default>;
+};
+
+&uart3 {
+	status = "okay";
+};
+
+&uart4 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <>;
+};
+
+&uart5 {
+	status = "okay";
+};
+
+&mac1 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii2_default &pinctrl_mdio2_default>;
+};
+
+&mac0 {
+	status = "okay";
+	use-ncsi;
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rmii1_default>;
+};
+
+&i2c0 {
+	multi-master;
+	general-call;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c1 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c2 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c3 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c4 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c5 {
+	bus-frequency = <1000000>;
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c6 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c7 {
+	multi-master;
+	#retries = <3>;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c9 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c11 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&i2c13 {
+	multi-master;
+	aspeed,dma-buf-size = <4095>;
+	aspeed,hw-timeout-ms = <300>;
+	status = "okay";
+};
+
+&gfx {
+	status = "okay";
+	memory-region = <&gfx_memory>;
+};
+
+&vuart {
+	status = "okay";
+};
+
+&pwm_tacho {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pwm0_default &pinctrl_pwm1_default
+			 &pinctrl_pwm2_default &pinctrl_pwm3_default
+			 &pinctrl_pwm4_default &pinctrl_pwm5_default
+			 &pinctrl_pwm6_default &pinctrl_pwm7_default>;
+
+	fan@0 {
+		reg = <0x00>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x00 0x01>;
+	};
+	fan@1 {
+		reg = <0x01>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x02 0x03>;
+	};
+	fan@2 {
+		reg = <0x02>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x04 0x05>;
+	};
+	fan@3 {
+		reg = <0x03>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x06 0x07>;
+	};
+	fan@4 {
+		reg = <0x04>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x08 0x09>;
+	};
+	fan@5 {
+		reg = <0x05>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x0A 0x0B>;
+	};
+	fan@6 {
+		reg = <0x06>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x0C 0x0D>;
+	};
+	fan@7 {
+		reg = <0x07>;
+		aspeed,fan-tach-ch = /bits/ 8 <0x0E 0x0F>;
+	};
+
+};
+
+&timer {
+/*
+ *	Available settings:
+ *	fttmr010,pwm-outputs = <5>, <6>, <7>, <8>;
+ *	pinctrl-0 = <&pinctrl_timer5_default &pinctrl_timer6_default
+ *			&pinctrl_timer7_default &pinctrl_timer8_default>;
+ */
+	fttmr010,pwm-outputs = <6>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_timer6_default>;
+	#pwm-cells = <3>;
+	status = "okay";
+};
+
+&video {
+	status = "okay";
+	memory-region = <&video_engine_memory>;
+};
+
+&vhub {
+	status = "okay";
+};
diff --git a/arch/arm/boot/dts/aspeed-g5.dtsi b/arch/arm/boot/dts/aspeed-g5.dtsi
index 5a09707d2b7f..0a28c33fb3a3 100644
--- a/arch/arm/boot/dts/aspeed-g5.dtsi
+++ b/arch/arm/boot/dts/aspeed-g5.dtsi
@@ -353,7 +353,7 @@
 				reg = <0x1e780200 0x0100>;
 				clocks = <&syscon ASPEED_CLK_APB>;
 				interrupt-controller;
-				bus-frequency = <12000000>;
+				bus-frequency = <1000000>;
 				pinctrl-names = "default";
 				pinctrl-0 = <&pinctrl_sgpm_default>;
 				status = "disabled";
@@ -462,34 +462,34 @@
 				#size-cells = <1>;
 				ranges = <0x0 0x1e789000 0x1000>;
 
-				kcs1: kcs@24 {
-					compatible = "aspeed,ast2500-kcs-bmc-v2";
-					reg = <0x24 0x1>, <0x30 0x1>, <0x3c 0x1>;
+				kcs1: kcs1@0 {
+					compatible = "aspeed,ast2500-kcs-bmc";
 					interrupts = <8>;
+					kcs_chan = <1>;
 					clocks = <&syscon ASPEED_CLK_GATE_LCLK>;
 					status = "disabled";
 				};
 
-				kcs2: kcs@28 {
-					compatible = "aspeed,ast2500-kcs-bmc-v2";
-					reg = <0x28 0x1>, <0x34 0x1>, <0x40 0x1>;
+				kcs2: kcs2@0 {
+					compatible = "aspeed,ast2500-kcs-bmc";
 					interrupts = <8>;
+					kcs_chan = <2>;
 					clocks = <&syscon ASPEED_CLK_GATE_LCLK>;
 					status = "disabled";
 				};
 
-				kcs3: kcs@2c {
-					compatible = "aspeed,ast2500-kcs-bmc-v2";
-					reg = <0x2c 0x1>, <0x38 0x1>, <0x44 0x1>;
+				kcs3: kcs3@0 {
+					compatible = "aspeed,ast2500-kcs-bmc";
 					interrupts = <8>;
+					kcs_chan = <3>;
 					clocks = <&syscon ASPEED_CLK_GATE_LCLK>;
 					status = "disabled";
 				};
 
-				kcs4: kcs@114 {
-					compatible = "aspeed,ast2500-kcs-bmc-v2";
-					reg = <0x114 0x1>, <0x118 0x1>, <0x11c 0x1>;
+				kcs4: kcs4@0 {
+					compatible = "aspeed,ast2500-kcs-bmc";
 					interrupts = <8>;
+					kcs_chan = <4>;
 					clocks = <&syscon ASPEED_CLK_GATE_LCLK>;
 					status = "disabled";
 				};
diff --git a/drivers/hwmon/Kconfig b/drivers/hwmon/Kconfig
index ff3e6fcfa893..e2af838bf256 100644
--- a/drivers/hwmon/Kconfig
+++ b/drivers/hwmon/Kconfig
@@ -1532,6 +1532,21 @@ config SENSORS_PECI_DIMMPOWER
 	  This driver can also be built as a module. If so, the module
 	  will be called peci-dimmpower.
 
+config SENSORS_PECI_PLATFORMPOWER
+	tristate "PECI platform power monitoring support"
+	depends on PECI
+	select MFD_INTEL_PECI_CLIENT
+	help
+	  If you say yes here you get support for the generic Intel PECI
+	  platformpower driver which provides average power of platform,
+	  current platform power limit, maximal and minimal power setting
+	  using the PECI Client Command Suite via the processor PECI client.
+	  Check Documentation/hwmon/peci-platformpower for details.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called peci-platformpower.
+
+
 source "drivers/hwmon/pmbus/Kconfig"
 
 config SENSORS_PWM_FAN
diff --git a/drivers/hwmon/Makefile b/drivers/hwmon/Makefile
index 67fbae3a22e6..fcbb4d3bdde7 100644
--- a/drivers/hwmon/Makefile
+++ b/drivers/hwmon/Makefile
@@ -159,6 +159,7 @@ obj-$(CONFIG_SENSORS_PECI_CPUTEMP)	+= peci-cputemp.o
 obj-$(CONFIG_SENSORS_PECI_DIMMTEMP)	+= peci-dimmtemp.o
 obj-$(CONFIG_SENSORS_PECI_CPUPOWER)	+= peci-cpupower.o
 obj-$(CONFIG_SENSORS_PECI_DIMMPOWER)	+= peci-dimmpower.o
+obj-$(CONFIG_SENSORS_PECI_PLATFORMPOWER)	+= peci-platformpower.o
 obj-$(CONFIG_SENSORS_POWR1220)  += powr1220.o
 obj-$(CONFIG_SENSORS_PWM_FAN)	+= pwm-fan.o
 obj-$(CONFIG_SENSORS_RASPBERRYPI_HWMON)	+= raspberrypi-hwmon.o
diff --git a/drivers/hwmon/peci-cputemp.c b/drivers/hwmon/peci-cputemp.c
index 8d5d579ccb1a..1763b53cc234 100644
--- a/drivers/hwmon/peci-cputemp.c
+++ b/drivers/hwmon/peci-cputemp.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 // Copyright (c) 2018-2019 Intel Corporation
 
+#include <linux/bitfield.h>
 #include <linux/hwmon.h>
 #include <linux/jiffies.h>
 #include <linux/mfd/intel-peci-client.h>
@@ -10,8 +11,11 @@
 #include "peci-hwmon.h"
 
 #define DEFAULT_CHANNEL_NUMS	5
-#define MODTEMP_CHANNEL_NUMS	CORE_MASK_BITS_MAX
+#define HBMTEMP_CHANNEL_NUMS	8
+#define HBM_PARAM_OFFSET	0xe0
+#define MODTEMP_CHANNEL_NUMS	(CORE_MASK_BITS_MAX + HBMTEMP_CHANNEL_NUMS)
 #define CPUTEMP_CHANNEL_NUMS	(DEFAULT_CHANNEL_NUMS + MODTEMP_CHANNEL_NUMS)
+#define HBM_ENABLED_MASK	GENMASK(30, 27)
 
 struct temp_group {
 	struct peci_sensor_data		die;
@@ -32,6 +36,7 @@ struct peci_cputemp {
 	const struct cpu_gen_info	*gen_info;
 	struct temp_group		temp;
 	u64				core_mask;
+	u64				hbm_mask;
 	u32				temp_config[CPUTEMP_CHANNEL_NUMS + 1];
 	uint				config_idx;
 	struct hwmon_channel_info	temp_info;
@@ -47,6 +52,7 @@ enum cputemp_channels {
 	channel_tthrottle,
 	channel_tjmax,
 	channel_core,
+	channel_hbm,
 };
 
 static const u32 config_table[] = {
@@ -69,6 +75,10 @@ static const u32 config_table[] = {
 
 	/* Core temperature - for all core channels */
 	HWMON_T_LABEL | HWMON_T_INPUT,
+
+	/* HBM temperature - for all high bandwidth memory channels */
+	HWMON_T_LABEL | HWMON_T_INPUT | HWMON_T_MAX | HWMON_T_CRIT |
+	HWMON_T_CRIT_HYST,
 };
 
 static const char *cputemp_label[DEFAULT_CHANNEL_NUMS] = {
@@ -241,14 +251,20 @@ static int get_module_temp(struct peci_cputemp *priv, int index)
 {
 	s32 module_dts_margin;
 	u8  pkg_cfg[4];
+	u16 param;
 	int ret;
 
 	if (!peci_sensor_need_update(&priv->temp.module[index]))
 		return 0;
 
+	if (index < CORE_MASK_BITS_MAX)
+		param = index;
+	else
+		param = index - CORE_MASK_BITS_MAX + HBM_PARAM_OFFSET;
+
 	ret = peci_client_read_package_config(priv->mgr,
 					      PECI_MBX_INDEX_MODULE_TEMP,
-					      index, pkg_cfg);
+					      param, pkg_cfg);
 	if (ret)
 		return ret;
 
@@ -367,12 +383,15 @@ static umode_t cputemp_is_visible(const void *data,
 				  u32 attr, int channel)
 {
 	const struct peci_cputemp *priv = data;
+	uint hbm_chan_offset = DEFAULT_CHANNEL_NUMS + CORE_MASK_BITS_MAX;
 
 	if (channel < ARRAY_SIZE(priv->temp_config) &&
 	    (priv->temp_config[channel] & BIT(attr)) &&
 	    (channel < DEFAULT_CHANNEL_NUMS ||
-	     (channel >= DEFAULT_CHANNEL_NUMS &&
-	      (priv->core_mask & BIT(channel - DEFAULT_CHANNEL_NUMS)))))
+	     (channel < hbm_chan_offset &&
+	      (priv->core_mask & BIT_ULL(channel - DEFAULT_CHANNEL_NUMS))) ||
+	     (channel >= hbm_chan_offset &&
+	      (priv->hbm_mask & BIT_ULL((channel - hbm_chan_offset) / 2)))))
 		return 0444;
 
 	return 0;
@@ -387,10 +406,44 @@ static const struct hwmon_ops cputemp_ops = {
 static int check_resolved_cores(struct peci_cputemp *priv)
 {
 	struct peci_rd_pci_cfg_local_msg msg;
+	struct peci_rd_end_pt_cfg_msg re_msg;
 	int ret;
 
 	/* Get the RESOLVED_CORES register value */
 	switch (priv->gen_info->model) {
+	case INTEL_FAM6_SAPPHIRERAPIDS:
+		re_msg.addr = priv->mgr->client->addr;
+		re_msg.msg_type = PECI_ENDPTCFG_TYPE_LOCAL_PCI;
+		re_msg.params.pci_cfg.seg = 0;
+		re_msg.params.pci_cfg.bus = 31;
+		re_msg.params.pci_cfg.device = 30;
+		re_msg.params.pci_cfg.function = 6;
+		re_msg.params.pci_cfg.reg = 0x84;
+		re_msg.rx_len = 4;
+
+		ret = peci_command(priv->mgr->client->adapter,
+				   PECI_CMD_RD_END_PT_CFG, &re_msg);
+		if (ret || re_msg.cc != PECI_DEV_CC_SUCCESS)
+			ret = -EAGAIN;
+		if (ret)
+			return ret;
+
+		priv->core_mask = le32_to_cpup((__le32 *)re_msg.data);
+		priv->core_mask <<= 32;
+
+		re_msg.params.pci_cfg.reg = 0x80;
+
+		ret = peci_command(priv->mgr->client->adapter,
+				   PECI_CMD_RD_END_PT_CFG, &re_msg);
+		if (ret || re_msg.cc != PECI_DEV_CC_SUCCESS)
+			ret = -EAGAIN;
+		if (ret) {
+			priv->core_mask = 0;
+			return ret;
+		}
+
+		priv->core_mask |= le32_to_cpup((__le32 *)re_msg.data);
+		break;
 	case INTEL_FAM6_ICELAKE_X:
 	case INTEL_FAM6_ICELAKE_XD:
 		msg.addr = priv->mgr->client->addr;
@@ -459,17 +512,26 @@ static int create_module_temp_label(struct peci_cputemp *priv, int idx)
 	if (!priv->module_temp_label[idx])
 		return -ENOMEM;
 
-	sprintf(priv->module_temp_label[idx], "Core %d", idx);
+	if (idx < CORE_MASK_BITS_MAX) {
+		sprintf(priv->module_temp_label[idx], "Core %d", idx);
+	} else {
+		int hbm_idx = idx - CORE_MASK_BITS_MAX;
+
+		sprintf(priv->module_temp_label[idx], "%s %d",
+			hbm_idx % 2 ? "HBM DRAM" : "HBM Logic",
+			hbm_idx / 2 + 1);
+	}
 
 	return 0;
 }
 
 static int create_module_temp_info(struct peci_cputemp *priv)
 {
+	u8 model = priv->gen_info->model;
 	int ret, i;
 
 	ret = check_resolved_cores(priv);
-	if (ret)
+	if (ret && model != INTEL_FAM6_SAPPHIRERAPIDS)
 		return ret;
 
 	priv->module_temp_label = devm_kzalloc(priv->dev,
@@ -479,10 +541,38 @@ static int create_module_temp_info(struct peci_cputemp *priv)
 	if (!priv->module_temp_label)
 		return -ENOMEM;
 
+	if (model == INTEL_FAM6_SAPPHIRERAPIDS) {
+		struct peci_rd_end_pt_cfg_msg re_msg;
+		u32 capid3_cfg;
+
+		re_msg.addr = priv->mgr->client->addr;
+		re_msg.rx_len = 4;
+		re_msg.msg_type = PECI_ENDPTCFG_TYPE_LOCAL_PCI;
+		re_msg.params.pci_cfg.seg = 0;
+		re_msg.params.pci_cfg.bus = 31;
+		re_msg.params.pci_cfg.device = 30;
+		re_msg.params.pci_cfg.function = 3;
+		re_msg.params.pci_cfg.reg = 0x90;
+
+		ret = peci_command(priv->mgr->client->adapter,
+				   PECI_CMD_RD_END_PT_CFG, &re_msg);
+		if (ret || re_msg.cc != PECI_DEV_CC_SUCCESS)
+			ret = -EAGAIN;
+		if (ret)
+			return ret;
+
+		capid3_cfg = le32_to_cpup((__le32 *)re_msg.data);
+		priv->hbm_mask = FIELD_GET(HBM_ENABLED_MASK, capid3_cfg);
+	}
+
 	for (i = 0; i < MODTEMP_CHANNEL_NUMS; i++) {
-		priv->temp_config[priv->config_idx++] = config_table[channel_core];
+		priv->temp_config[priv->config_idx++] = i < CORE_MASK_BITS_MAX ?
+			config_table[channel_core] : config_table[channel_hbm];
 
-		if (i < priv->gen_info->core_mask_bits && priv->core_mask & BIT(i)) {
+		if ((i < priv->gen_info->core_mask_bits &&
+		     priv->core_mask & BIT_ULL(i)) ||
+		    (i >= CORE_MASK_BITS_MAX &&
+		     priv->hbm_mask & BIT_ULL((i - CORE_MASK_BITS_MAX) / 2))) {
 			ret = create_module_temp_label(priv, i);
 			if (ret)
 				return ret;
diff --git a/drivers/hwmon/peci-dimmtemp.c b/drivers/hwmon/peci-dimmtemp.c
index 8b3a265355e0..bc44428c997b 100644
--- a/drivers/hwmon/peci-dimmtemp.c
+++ b/drivers/hwmon/peci-dimmtemp.c
@@ -42,6 +42,7 @@ static const u8 support_model[] = {
 	INTEL_FAM6_SKYLAKE_XD,
 	INTEL_FAM6_ICELAKE_X,
 	INTEL_FAM6_ICELAKE_XD,
+	INTEL_FAM6_SAPPHIRERAPIDS,
 };
 
 static inline int read_ddr_dimm_temp_config(struct peci_dimmtemp *priv,
@@ -73,6 +74,71 @@ static int get_dimm_temp(struct peci_dimmtemp *priv, int dimm_no)
 	priv->temp[dimm_no].value = cfg_data[dimm_order] * 1000;
 
 	switch (priv->gen_info->model) {
+	case INTEL_FAM6_SAPPHIRERAPIDS:
+		re_msg.addr = priv->mgr->client->addr;
+		re_msg.rx_len = 4;
+		re_msg.msg_type = PECI_ENDPTCFG_TYPE_LOCAL_PCI;
+		re_msg.params.pci_cfg.seg = 0;
+		re_msg.params.pci_cfg.bus = 30;
+		re_msg.params.pci_cfg.device = 0;
+		re_msg.params.pci_cfg.function = 2;
+		re_msg.params.pci_cfg.reg = 0xd4;
+
+		ret = peci_command(priv->mgr->client->adapter,
+				   PECI_CMD_RD_END_PT_CFG, &re_msg);
+		if (ret || re_msg.cc != PECI_DEV_CC_SUCCESS ||
+		    !(re_msg.data[3] & BIT(7))) {
+			/* Use default or previous value */
+			ret = 0;
+			break;
+		}
+
+		re_msg.params.pci_cfg.reg = 0xd0;
+
+		ret = peci_command(priv->mgr->client->adapter,
+				   PECI_CMD_RD_END_PT_CFG, &re_msg);
+		if (ret || re_msg.cc != PECI_DEV_CC_SUCCESS) {
+			/* Use default or previous value */
+			ret = 0;
+			break;
+		}
+
+		cpu_seg = re_msg.data[2];
+		cpu_bus = re_msg.data[0];
+
+		re_msg.msg_type = PECI_ENDPTCFG_TYPE_MMIO;
+		re_msg.params.mmio.seg = cpu_seg;
+		re_msg.params.mmio.bus = cpu_bus;
+		/*
+		 * Device 26, Offset 219a8: IMC 0 channel 0 -> rank 0
+		 * Device 26, Offset 299a8: IMC 0 channel 1 -> rank 1
+		 * Device 27, Offset 219a8: IMC 1 channel 0 -> rank 2
+		 * Device 27, Offset 299a8: IMC 1 channel 1 -> rank 3
+		 * Device 28, Offset 219a8: IMC 2 channel 0 -> rank 4
+		 * Device 28, Offset 299a8: IMC 2 channel 1 -> rank 5
+		 * Device 29, Offset 219a8: IMC 3 channel 0 -> rank 6
+		 * Device 29, Offset 299a8: IMC 3 channel 1 -> rank 7
+		 */
+		re_msg.params.mmio.device = 0x1a + chan_rank / 2;
+		re_msg.params.mmio.function = 0;
+		re_msg.params.mmio.bar = 0;
+		re_msg.params.mmio.addr_type = PECI_ENDPTCFG_ADDR_TYPE_MMIO_Q;
+		re_msg.params.mmio.offset = 0x219a8 + dimm_order * 4;
+		if (chan_rank % 2)
+			re_msg.params.mmio.offset += 0x8000;
+
+		ret = peci_command(priv->mgr->client->adapter,
+				   PECI_CMD_RD_END_PT_CFG, &re_msg);
+		if (ret || re_msg.cc != PECI_DEV_CC_SUCCESS ||
+		    re_msg.data[1] == 0 || re_msg.data[2] == 0) {
+			/* Use default or previous value */
+			ret = 0;
+			break;
+		}
+
+		priv->temp_max[dimm_no] = re_msg.data[1] * 1000;
+		priv->temp_crit[dimm_no] = re_msg.data[2] * 1000;
+		break;
 	case INTEL_FAM6_ICELAKE_X:
 	case INTEL_FAM6_ICELAKE_XD:
 		re_msg.addr = priv->mgr->client->addr;
diff --git a/drivers/hwmon/peci-platformpower.c b/drivers/hwmon/peci-platformpower.c
new file mode 100644
index 000000000000..90c355c61238
--- /dev/null
+++ b/drivers/hwmon/peci-platformpower.c
@@ -0,0 +1,910 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020-2020 Intel Corporation
+
+#include <linux/hwmon.h>
+#include <linux/jiffies.h>
+#include <linux/mfd/intel-peci-client.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+#include <linux/platform_device.h>
+#include "peci-hwmon.h"
+
+enum PECI_PLATFORMPOWER_POWER_SENSOR_TYPES {
+	PECI_PLATFORMPOWER_SENSOR_TYPE_POWER = 0,
+	PECI_PLATFORMPOWER_SENSOR_TYPE_ENERGY,
+	PECI_PLATFORMPOWER_SENSOR_TYPES_COUNT,
+};
+
+#define PECI_PLATFORMPOWER_POWER_CHANNEL_COUNT		1 /* Supported channels number. */
+#define PECI_PLATFORMPOWER_ENERGY_CHANNEL_COUNT	1 /* Supported channels number. */
+
+#define PECI_PLATFORMPOWER_POWER_SENSOR_COUNT	4 /* Supported sensors number. */
+#define PECI_PLATFORMPOWER_ENERGY_SENSOR_COUNT	1 /* Supported sensors number. */
+
+#define PECI_PLATFORMPOWER_ENERGY_UNIT	0
+
+/**
+ * union peci_platform_power_info_low - Platform and Package Power Info (Low)
+ * PCS. This PCS coresponds to the MSR@665h - PLATFORM_POWER_INFO, bits [31:0]
+ * Accessing over PECI: PCS=0x1C, parameter=0x00FE
+ * @value: PCS register value
+ * @bits:	PCS register bits
+ *		@max_ppl1:	Bits [15:0] - Max PP L1 value
+ *		@rsvd0:		Bits [16:16]
+ *		@min_ppl1:	Bits [31:17] - Min PP L1 value
+ */
+union peci_platform_power_info_low {
+	u32 value;
+	struct {
+		u32 max_ppl1        : 16;
+		u32 rsvd0           : 1;
+		u32 min_ppl1        : 15;
+	} __attribute__((__packed__)) bits;
+} __attribute__((__packed__));
+
+static_assert(sizeof(union peci_platform_power_info_low) ==
+		     PECI_PCS_REGISTER_SIZE);
+
+/**
+ * union peci_platform_power_limit_high - Platform Power Limit 2 PCS
+ * This PCS coresponds to the MSR@65Ch - PLATFORM_POWER_LIMIT_SRVR, bits [63:32]
+ * Accessing over PECI: PCS=0x3B, Parameter=0x0000
+ * @value: PCS register value
+ * @bits:	PCS register bits
+ *		@pwr_lim_2:	Bits [16:0] - Power Limit 2
+ *		@pwr_lim_2_en:	Bits [17:17] - Power Limit 2 Enable
+ *		@pwr_clmp_lim_2:Bits [18:18] - Critical Power Clamp 2
+ *		@pwr_lim_2_time:Bits [25:19] - Power Limit 2 Time Window
+ *		@rsvd0:		Bits [31:26]
+ */
+union peci_platform_power_limit_high {
+	u32 value;
+	struct {
+		u32 pwr_lim_2       : 17;
+		u32 pwr_lim_2_en    : 1;
+		u32 pwr_clmp_lim_2  : 1;
+		u32 pwr_lim_2_time  : 7;
+		u32 rsvd0           : 6;
+	} __attribute__((__packed__)) bits;
+} __attribute__((__packed__));
+
+static_assert(sizeof(union peci_platform_power_limit_high) ==
+	      PECI_PCS_REGISTER_SIZE);
+
+/**
+ * union peci_platform_power_limit_low - Platform Power Limit 1 PCS
+ * This PCS coresponds to the MSR@65Ch - PLATFORM_POWER_LIMIT_SRVR, bits [31:0]
+ * Accessing over PECI: PCS=0x3A, Parameter=0x0000
+ * @value: PCS register value
+ * @bits:	PCS register bits
+ *		@pwr_lim_1:	Bits [16:0] - Power Limit 1
+ *		@pwr_lim_1_en:	Bits [17:17] - Power Limit 1 Enable
+ *		@pwr_clmp_lim_1:Bits [18:18] - Critical Power Clamp 1
+ *		@pwr_lim_1_time:Bits [25:19] - Power Limit 1 Time Window
+ *		@rsvd0:		Bits [31:26]
+ */
+union peci_platform_power_limit_low {
+	u32 value;
+	struct {
+		u32 pwr_lim_1       : 17;
+		u32 pwr_lim_1_en    : 1;
+		u32 pwr_clmp_lim_1  : 1;
+		u32 pwr_lim_1_time  : 7;
+		u32 rsvd0           : 6;
+	} __attribute__((__packed__)) bits;
+} __attribute__((__packed__));
+
+static_assert(sizeof(union peci_platform_power_limit_low) ==
+		     PECI_PCS_REGISTER_SIZE);
+
+struct peci_platformpower {
+	struct device *dev;
+	struct peci_client_manager *mgr;
+	char name[PECI_NAME_SIZE];
+	const struct cpu_gen_info *gen_info;
+	u32 power_config[PECI_PLATFORMPOWER_POWER_CHANNEL_COUNT + 1];
+	u32 energy_config[PECI_PLATFORMPOWER_ENERGY_CHANNEL_COUNT + 1];
+
+	struct hwmon_channel_info power_info;
+	struct hwmon_channel_info energy_info;
+	const struct hwmon_channel_info *info[PECI_PLATFORMPOWER_SENSOR_TYPES_COUNT + 1];
+	struct hwmon_chip_info chip;
+
+	struct peci_sensor_data
+		power_sensor_data_list[PECI_PLATFORMPOWER_POWER_CHANNEL_COUNT]
+				      [PECI_PLATFORMPOWER_POWER_SENSOR_COUNT];
+	struct peci_sensor_data
+		energy_sensor_data_list[PECI_PLATFORMPOWER_ENERGY_CHANNEL_COUNT]
+				       [PECI_PLATFORMPOWER_ENERGY_SENSOR_COUNT];
+
+	/* Below structs are not exposed to any sensor directly */
+	struct peci_sensor_data energy_cache; /* used to limit PECI communication */
+	struct peci_sensor_data power_sensor_prev_energy;
+	struct peci_sensor_data energy_sensor_prev_energy;
+
+	union peci_pkg_power_sku_unit units;
+	bool units_valid;
+
+	u32 ppl1_time_window;
+	u32 ppl2_time_window;
+	bool ppl_time_windows_valid;
+};
+
+static const u8 peci_platformpower_models[] = {
+	INTEL_FAM6_SAPPHIRERAPIDS,
+};
+
+static const char
+*peci_platformpower_labels[PECI_PLATFORMPOWER_SENSOR_TYPES_COUNT] = {
+	"platform power",
+	"platform energy",
+};
+
+/**
+ * peci_platformpower_read_platf_pwr_info_low - read PCS Platform Power Info low
+ * @peci_mgr: PECI client manager handle
+ * @reg: Pointer to the variable read value is going to be put
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static inline int
+peci_platformpower_read_platf_pwr_info_low(struct peci_client_manager *peci_mgr,
+					   union peci_platform_power_info_low
+					   *reg)
+{
+	return peci_pcs_read(peci_mgr, PECI_MBX_INDEX_TDP,
+			     PECI_PKG_ID_PLATFORM, &reg->value);
+}
+
+/**
+ * peci_platformpower_read_platf_pwr_lim_low - read PCS Platform Power Limit Low
+ * @peci_mgr: PECI client manager handle
+ * @reg: Pointer to the variable read value is going to be put
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static inline int
+peci_platformpower_read_platf_pwr_lim_low(struct peci_client_manager *peci_mgr,
+					  union peci_platform_power_limit_low
+					  *reg)
+{
+	return peci_pcs_read(peci_mgr, PECI_MBX_INDEX_PKG_PSYS_PWR_LIM1,
+			     PECI_PCS_PARAM_ZERO, &reg->value);
+}
+
+/**
+ * peci_platformpower_get_energy_counter - get actual energy counter value
+ * @priv: Pointer to the peci platformpower context
+ * @sensor_data: Sensor data
+ * @update_interval: time in jiffies needs to elapse to read sensor again
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static int
+peci_platformpower_get_energy_counter(struct peci_platformpower *priv,
+				      struct peci_sensor_data *sensor_data,
+				      ulong update_interval)
+{
+	int ret;
+
+	if (!peci_sensor_need_update_with_time(sensor_data,
+					       update_interval)) {
+		dev_dbg(priv->dev, "skip reading package energy over peci\n");
+		return 0;
+	}
+
+	ret = peci_pcs_read(priv->mgr, PECI_MBX_INDEX_ENERGY_COUNTER,
+			    PECI_PKG_ID_PLATFORM, &sensor_data->uvalue);
+
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read package energy\n");
+		return ret;
+	}
+
+	peci_sensor_mark_updated(sensor_data);
+
+	dev_dbg(priv->dev,
+		"energy counter updated %duJ, jif %lu, HZ is %d jiffies\n",
+		sensor_data->uvalue, sensor_data->last_updated, HZ);
+
+	return ret;
+}
+
+/**
+ * peci_platformpower_get_average_power - get avarage platform power
+ * @ctx: Pointer to the driver context
+ * @sensor_conf: Sensor configuration
+ * @sensor_data: Sensor data
+ * @val: Pointer to the variable read value is going to be put, in milliwatts
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static int
+peci_platformpower_get_average_power(void *ctx,
+				     struct peci_sensor_conf *sensor_conf,
+				     struct peci_sensor_data *sensor_data)
+{
+	struct peci_platformpower *priv = (struct peci_platformpower *)ctx;
+	int ret;
+
+	if (!peci_sensor_need_update_with_time(sensor_data,
+					       sensor_conf->update_interval)) {
+		dev_dbg(priv->dev,
+			"skip generating new power value %dmW jif %lu\n",
+			sensor_data->value, jiffies);
+		return 0;
+	}
+
+	ret = peci_platformpower_get_energy_counter(priv, &priv->energy_cache,
+						    sensor_conf->update_interval);
+	if (ret) {
+		dev_dbg(priv->dev, "cannot update energy counter\n");
+		return ret;
+	}
+
+	ret = peci_pcs_calc_pwr_from_eng(priv->dev,
+					 &priv->power_sensor_prev_energy,
+					 &priv->energy_cache,
+					 PECI_PLATFORMPOWER_ENERGY_UNIT,
+					 &sensor_data->value);
+	if (ret) {
+		dev_dbg(priv->dev, "power calculation failed\n");
+		return ret;
+	}
+
+	peci_sensor_mark_updated_with_time(sensor_data,
+					   priv->energy_cache.last_updated);
+
+	dev_dbg(priv->dev, "average power %dmW, jif %lu, HZ is %d jiffies\n",
+		sensor_data->value, sensor_data->last_updated, HZ);
+
+	return ret;
+}
+
+/**
+ * peci_platformpower_get_power_limit - get current platform power limit
+ * @ctx: Pointer to the driver context
+ * @sensor_conf: Sensor configuration
+ * @sensor_data: Sensor data
+ * @val: Pointer to the variable read value is going to be put, in milliwatts
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static int
+peci_platformpower_get_power_limit(void *ctx,
+				   struct peci_sensor_conf *sensor_conf,
+				   struct peci_sensor_data *sensor_data)
+{
+	struct peci_platformpower *priv = (struct peci_platformpower *)ctx;
+	union peci_platform_power_limit_low power_limit;
+	int ret;
+
+	/*
+	 * Check whether need to update reading. If not just return cached
+	 * value.
+	 */
+	if (!peci_sensor_need_update_with_time(sensor_data,
+					       sensor_conf->update_interval)) {
+		dev_dbg(priv->dev, "skip reading peci, power limit %dmW\n",
+			sensor_data->value);
+		return 0;
+	}
+
+	/*
+	 * Read units to the cache. Units are needed to convert power values
+	 * correctly. Units are read from CPU only once.
+	 */
+	ret = peci_pcs_get_units(priv->mgr, &priv->units, &priv->units_valid);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read units\n");
+		return ret;
+	}
+
+	/* Read power limit 1 (PPL1). */
+	ret = peci_platformpower_read_platf_pwr_lim_low(priv->mgr,
+							&power_limit);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read power limit 1 (PPL1)\n");
+		return ret;
+	}
+
+	peci_sensor_mark_updated(sensor_data);
+	sensor_data->value = peci_pcs_xn_to_munits(power_limit.bits.pwr_lim_1,
+						   priv->units.bits.pwr_unit);
+
+	dev_dbg(priv->dev, "raw power limit %u, unit %u, power limit %d\n",
+		power_limit.bits.pwr_lim_1, priv->units.bits.pwr_unit,
+		sensor_data->value);
+
+	return ret;
+}
+
+/**
+ * peci_platformpower_set_power_limit - set platform power limit
+ * @ctx: Pointer to the driver context
+ * @sensor_conf: Sensor configuration
+ * @sensor_data: Sensor data
+ * @val: Power limit to set, in milliwatts
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static int
+peci_platformpower_set_power_limit(void *ctx,
+				   struct peci_sensor_conf *sensor_conf,
+				   struct peci_sensor_data *sensor_data,
+				   s32 val)
+{
+	struct peci_platformpower *priv = (struct peci_platformpower *)ctx;
+	union peci_platform_power_limit_high power_limit_high;
+	union peci_platform_power_limit_low power_limit_low;
+	int ret;
+
+	/*
+	 * Read units to the cache. Units are needed to convert power values
+	 * correctly. Units are read from CPU only once.
+	 */
+	ret = peci_pcs_get_units(priv->mgr, &priv->units, &priv->units_valid);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read units\n");
+		return ret;
+	}
+
+	/* Read power limit 1 (PPL1). */
+	ret = peci_platformpower_read_platf_pwr_lim_low(priv->mgr,
+							&power_limit_low);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read power limit 1 (PPL1)\n");
+		return ret;
+	}
+
+	/* Read power limit 2 (PPL2). */
+	ret = peci_pcs_read(priv->mgr, PECI_MBX_INDEX_PKG_PSYS_PWR_LIM2,
+			    PECI_PCS_PARAM_ZERO, &power_limit_high.value);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read power limit 2 (PPL2)\n");
+		return ret;
+	}
+
+	/* Calculate PPL time windows if needed */
+	if (!priv->ppl_time_windows_valid) {
+		priv->ppl1_time_window =
+			peci_pcs_calc_plxy_time_window(peci_pcs_munits_to_xn
+				(PECI_PCS_PPL1_TIME_WINDOW,
+				priv->units.bits.tim_unit));
+		priv->ppl2_time_window =
+			peci_pcs_calc_plxy_time_window(peci_pcs_munits_to_xn
+				(PECI_PCS_PPL2_TIME_WINDOW,
+				priv->units.bits.tim_unit));
+		priv->ppl_time_windows_valid = true;
+	}
+
+	/* Enable or disable power limitation */
+	if (val > 0) {
+		/* Calculate and enable PPL1 */
+		power_limit_low.bits.pwr_lim_1 =
+			peci_pcs_munits_to_xn(val, priv->units.bits.pwr_unit);
+		power_limit_low.bits.pwr_lim_1_en = 1u;
+		power_limit_low.bits.pwr_clmp_lim_1 = 1u;
+		power_limit_low.bits.pwr_lim_1_time = priv->ppl1_time_window;
+
+		/* Calculate and enable PPL2 */
+		power_limit_high.bits.pwr_lim_2 =
+			peci_pcs_munits_to_xn(PECI_PCS_PPL1_TO_PPL2(val),
+					      priv->units.bits.pwr_unit);
+		power_limit_high.bits.pwr_lim_2_en = 1u;
+		power_limit_high.bits.pwr_clmp_lim_2 = 1u;
+		power_limit_high.bits.pwr_lim_2_time = priv->ppl2_time_window;
+	} else {
+		/* Disable power limitation */
+		power_limit_low.bits.pwr_lim_1 = 0u;
+		power_limit_low.bits.pwr_lim_1_en = 0u;
+		power_limit_low.bits.pwr_clmp_lim_1 = 0u;
+		power_limit_low.bits.pwr_lim_1_time = 0u;
+		power_limit_high.bits.pwr_lim_2 = 0u;
+		power_limit_high.bits.pwr_lim_2_en = 0u;
+		power_limit_high.bits.pwr_clmp_lim_2 = 0u;
+		power_limit_high.bits.pwr_lim_2_time = 0u;
+	}
+
+	/* Write calculated PPL1 to the CPU */
+	ret = peci_pcs_write(priv->mgr, PECI_MBX_INDEX_PKG_PSYS_PWR_LIM1,
+			     PECI_PCS_PARAM_ZERO, power_limit_low.value);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to write power limit 1\n");
+		return ret;
+	}
+
+	/* Write calculated PPL2 to the CPU */
+	ret = peci_pcs_write(priv->mgr, PECI_MBX_INDEX_PKG_PSYS_PWR_LIM2,
+			     PECI_PCS_PARAM_ZERO, power_limit_high.value);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to write power limit 2\n");
+		return ret;
+	}
+
+	dev_dbg(priv->dev, "power limit %d, unit %u, raw power limit 1 %u,\n",
+		val, priv->units.bits.pwr_unit, power_limit_low.bits.pwr_lim_1);
+
+	return ret;
+}
+
+/**
+ * peci_platformpower_read_max_power - get maximum value for the platform power
+ * limit
+ * @ctx: Pointer to the driver context
+ * @sensor_conf: Sensor configuration
+ * @sensor_data: Sensor data
+ * @val: Pointer to the variable read value is going to be put, in milliwatts
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static int
+peci_platformpower_read_max_power(void *ctx,
+				  struct peci_sensor_conf *sensor_conf,
+				  struct peci_sensor_data *sensor_data)
+{
+	struct peci_platformpower *priv = (struct peci_platformpower *)ctx;
+	union peci_platform_power_info_low power_info;
+	int ret;
+
+	/*
+	 * Check whether need to update reading. If not just return cached
+	 * value.
+	 */
+	if (!peci_sensor_need_update_with_time(sensor_data,
+					       sensor_conf->update_interval)) {
+		dev_dbg(priv->dev, "skip reading peci, max power %dmW\n",
+			sensor_data->value);
+		return 0;
+	}
+
+	/*
+	 * Read units to the cache. Units are needed to convert power values
+	 * correctly. Units are read from CPU only once.
+	 */
+	ret = peci_pcs_get_units(priv->mgr, &priv->units, &priv->units_valid);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read units\n");
+		return ret;
+	}
+
+	/* Read platform power info. */
+	ret = peci_platformpower_read_platf_pwr_info_low(priv->mgr,
+							 &power_info);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read platform info\n");
+		return ret;
+	}
+
+	peci_sensor_mark_updated(sensor_data);
+	sensor_data->value = peci_pcs_xn_to_munits(power_info.bits.max_ppl1,
+						   priv->units.bits.pwr_unit);
+
+	dev_dbg(priv->dev, "raw max power %u, unit %u, max power %dmW\n",
+		power_info.bits.max_ppl1, priv->units.bits.pwr_unit,
+		sensor_data->value);
+
+	return ret;
+}
+
+/**
+ * peci_platformpower_read_min_power - get minimal value for the platform power
+ * limit
+ * @ctx: Pointer to the driver context
+ * @sensor_conf: Sensor configuration
+ * @sensor_data: Sensor data
+ * @val: Pointer to the variable read value is going to be put, in milliwatts
+ *
+ * Return: 0 if succeeded, other values in case an error.
+ */
+static int
+peci_platformpower_read_min_power(void *ctx,
+				  struct peci_sensor_conf *sensor_conf,
+				  struct peci_sensor_data *sensor_data)
+{
+	struct peci_platformpower *priv = (struct peci_platformpower *)ctx;
+	union peci_platform_power_info_low power_info;
+	int ret;
+
+	/*
+	 * Check whether need to update reading. If not just return cached
+	 * value.
+	 */
+	if (!peci_sensor_need_update_with_time(sensor_data,
+					       sensor_conf->update_interval)) {
+		dev_dbg(priv->dev, "skip reading peci, min power %dmW\n",
+			sensor_data->value);
+		return 0;
+	}
+
+	/*
+	 * Read units to the cache. Units are needed to convert power values
+	 * correctly. Units are read from CPU only once.
+	 */
+	ret = peci_pcs_get_units(priv->mgr, &priv->units, &priv->units_valid);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read units\n");
+		return ret;
+	}
+
+	/* Read platform power info. */
+	ret = peci_platformpower_read_platf_pwr_info_low(priv->mgr,
+							 &power_info);
+	if (ret) {
+		dev_dbg(priv->dev, "not able to read platform info\n");
+		return ret;
+	}
+
+	peci_sensor_mark_updated(sensor_data);
+	sensor_data->value = peci_pcs_xn_to_munits(power_info.bits.min_ppl1,
+						   priv->units.bits.pwr_unit);
+
+	dev_dbg(priv->dev, "raw min power %u, unit %u, min power %dmW\n",
+		power_info.bits.min_ppl1, priv->units.bits.pwr_unit,
+		sensor_data->value);
+
+	return ret;
+}
+
+static int
+peci_platformpower_read_energy(void *ctx, struct peci_sensor_conf *sensor_conf,
+			       struct peci_sensor_data *sensor_data)
+{
+	struct peci_platformpower *priv = (struct peci_platformpower *)ctx;
+	int ret;
+
+	if (!peci_sensor_need_update_with_time(sensor_data,
+					       sensor_conf->update_interval)) {
+		dev_dbg(priv->dev,
+			"skip generating new energy value %duJ jif %lu\n",
+			sensor_data->value, jiffies);
+		return 0;
+	}
+
+	ret = peci_platformpower_get_energy_counter(priv, &priv->energy_cache,
+						    sensor_conf->update_interval);
+	if (ret) {
+		dev_dbg(priv->dev, "cannot update energy counter\n");
+		return ret;
+	}
+
+	ret = peci_pcs_calc_acc_eng(priv->dev,
+				    &priv->energy_sensor_prev_energy,
+				    &priv->energy_cache,
+				    PECI_PLATFORMPOWER_ENERGY_UNIT,
+				    &sensor_data->uvalue);
+
+	if (ret) {
+		dev_dbg(priv->dev, "cumulative energy calculation failed\n");
+		return ret;
+	}
+
+	peci_sensor_mark_updated_with_time(sensor_data,
+					   priv->energy_cache.last_updated);
+
+	dev_dbg(priv->dev, "energy %duJ, jif %lu, HZ is %d jiffies\n",
+		sensor_data->uvalue, sensor_data->last_updated, HZ);
+
+	return 0;
+}
+
+static struct peci_sensor_conf
+peci_platformpower_power_cfg[PECI_PLATFORMPOWER_POWER_CHANNEL_COUNT]
+			    [PECI_PLATFORMPOWER_POWER_SENSOR_COUNT] = {
+	/* Channel 0  - Power */
+	{
+		{
+			.attribute = hwmon_power_average,
+			.config = HWMON_P_AVERAGE,
+			.update_interval = UPDATE_INTERVAL_100MS,
+			.read = peci_platformpower_get_average_power,
+			.write = NULL,
+		},
+		{
+			.attribute = hwmon_power_cap,
+			.config = HWMON_P_CAP,
+			.update_interval = UPDATE_INTERVAL_DEFAULT,
+			.read = peci_platformpower_get_power_limit,
+			.write = peci_platformpower_set_power_limit,
+		},
+		{
+			.attribute = hwmon_power_cap_max,
+			.config = HWMON_P_CAP_MAX,
+			.update_interval = UPDATE_INTERVAL_10S,
+			.read = peci_platformpower_read_max_power,
+			.write = NULL,
+		},
+		{
+			.attribute = hwmon_power_cap_min,
+			.config = HWMON_P_CAP_MIN,
+			.update_interval = UPDATE_INTERVAL_10S,
+			.read = peci_platformpower_read_min_power,
+			.write = NULL,
+		},
+	},
+};
+
+static struct peci_sensor_conf
+peci_platformpower_energy_cfg[PECI_PLATFORMPOWER_ENERGY_CHANNEL_COUNT]
+			     [PECI_PLATFORMPOWER_ENERGY_SENSOR_COUNT] = {
+	/* Channel 0  - Energy */
+	{
+		{
+			.attribute = hwmon_energy_input,
+			.config = HWMON_E_INPUT,
+			.update_interval = UPDATE_INTERVAL_100MS,
+			.read = peci_platformpower_read_energy,
+			.write = NULL,
+		},
+	}
+};
+
+static bool
+peci_platformpower_is_channel_valid(enum hwmon_sensor_types type,
+				    int channel)
+{
+	if ((type == hwmon_power && channel < PECI_PLATFORMPOWER_POWER_CHANNEL_COUNT) ||
+	    (type == hwmon_energy && channel < PECI_PLATFORMPOWER_ENERGY_CHANNEL_COUNT))
+		return true;
+
+	return false;
+}
+
+static int
+peci_platformpower_read_string(struct device *dev, enum hwmon_sensor_types type,
+			       u32 attr, int channel, const char **str)
+{
+	if (!peci_platformpower_is_channel_valid(type, channel))
+		return -EOPNOTSUPP;
+
+	switch (attr) {
+	case hwmon_power_label:
+		*str = peci_platformpower_labels[PECI_PLATFORMPOWER_SENSOR_TYPE_POWER];
+		break;
+	case hwmon_energy_label:
+		*str = peci_platformpower_labels[PECI_PLATFORMPOWER_SENSOR_TYPE_ENERGY];
+		break;
+	default:
+		return -EOPNOTSUPP;
+	}
+
+	return 0;
+}
+
+static int
+peci_platformpower_read(struct device *dev, enum hwmon_sensor_types type,
+			u32 attr, int channel, long *val)
+{
+	struct peci_platformpower *priv = dev_get_drvdata(dev);
+	struct peci_sensor_conf *sensor_conf;
+	struct peci_sensor_data *sensor_data;
+	int ret;
+
+	if (!priv || !val)
+		return -EINVAL;
+
+	if (!peci_platformpower_is_channel_valid(type, channel))
+		return -EOPNOTSUPP;
+
+	/* Get sensor configuration and data. */
+	switch (type) {
+	case hwmon_power:
+		ret = peci_sensor_get_ctx(attr, peci_platformpower_power_cfg[channel],
+					  &sensor_conf,
+					  priv->power_sensor_data_list[channel],
+					  &sensor_data,
+					  ARRAY_SIZE(peci_platformpower_power_cfg[channel]));
+		break;
+	case hwmon_energy:
+		ret = peci_sensor_get_ctx(attr, peci_platformpower_energy_cfg[channel],
+					  &sensor_conf,
+					  priv->energy_sensor_data_list[channel],
+					  &sensor_data,
+					  ARRAY_SIZE(peci_platformpower_energy_cfg[channel]));
+		break;
+	default:
+		ret = -EOPNOTSUPP;
+	}
+
+	if (ret)
+		return ret;
+
+	if (sensor_conf->read) {
+		ret = sensor_conf->read(priv, sensor_conf, sensor_data);
+		if (!ret)
+			*val = (long)sensor_data->value;
+	} else {
+		ret = -EOPNOTSUPP;
+	}
+
+	return ret;
+}
+
+static int
+peci_platformpower_write(struct device *dev, enum hwmon_sensor_types type,
+			 u32 attr, int channel, long val)
+{
+	struct peci_platformpower *priv = dev_get_drvdata(dev);
+	struct peci_sensor_conf *sensor_conf;
+	struct peci_sensor_data *sensor_data;
+	int ret;
+
+	if (!priv)
+		return -EINVAL;
+
+	if (!peci_platformpower_is_channel_valid(type, channel))
+		return -EOPNOTSUPP;
+
+	/* Get sensor configuration and data. */
+	switch (type) {
+	case hwmon_power:
+		ret = peci_sensor_get_ctx(attr, peci_platformpower_power_cfg[channel],
+					  &sensor_conf,
+					  priv->power_sensor_data_list[channel],
+					  &sensor_data,
+					  ARRAY_SIZE(peci_platformpower_power_cfg[channel]));
+		break;
+	case hwmon_energy:
+		ret = peci_sensor_get_ctx(attr, peci_platformpower_energy_cfg[channel],
+					  &sensor_conf,
+					  priv->energy_sensor_data_list[channel],
+					  &sensor_data,
+					  ARRAY_SIZE(peci_platformpower_energy_cfg[channel]));
+		break;
+	default:
+		ret = -EOPNOTSUPP;
+	}
+
+	if (ret)
+		return ret;
+
+	if (sensor_conf->write) {
+		ret = sensor_conf->write(priv, sensor_conf, sensor_data,
+					 (s32)val);
+	} else {
+		ret = -EOPNOTSUPP;
+	}
+
+	return ret;
+}
+
+static umode_t
+peci_platformpower_is_visible(const void *data, enum hwmon_sensor_types type,
+			      u32 attr, int channel)
+{
+	struct peci_sensor_conf *sensor_conf;
+	umode_t mode = 0;
+	int ret;
+
+	if (!peci_platformpower_is_channel_valid(type, channel))
+		return mode;
+
+	/* If it is about label add read access only. */
+	if (attr == hwmon_power_label || attr == hwmon_energy_label)
+		return 0444;
+
+	/* Get sensor configuration and data. */
+	switch (type) {
+	case hwmon_power:
+		ret = peci_sensor_get_ctx(attr, peci_platformpower_power_cfg[channel],
+					  &sensor_conf, NULL, NULL,
+					  ARRAY_SIZE(peci_platformpower_power_cfg[channel]));
+		break;
+	case hwmon_energy:
+		ret = peci_sensor_get_ctx(attr, peci_platformpower_energy_cfg[channel],
+					  &sensor_conf, NULL, NULL,
+					  ARRAY_SIZE(peci_platformpower_energy_cfg[channel]));
+		break;
+	default:
+		return mode;
+	}
+
+	if (!ret) {
+		if (sensor_conf->read)
+			mode |= 0444;
+		if (sensor_conf->write)
+			mode |= 0200;
+	}
+
+	return mode;
+}
+
+static const struct hwmon_ops peci_platformpower_ops = {
+	.is_visible = peci_platformpower_is_visible,
+	.read_string = peci_platformpower_read_string,
+	.read = peci_platformpower_read,
+	.write = peci_platformpower_write,
+};
+
+static int peci_platformpower_probe(struct platform_device *pdev)
+{
+	struct peci_client_manager *mgr = dev_get_drvdata(pdev->dev.parent);
+	struct device *dev = &pdev->dev;
+	struct peci_platformpower *priv;
+	struct device *hwmon_dev;
+	u32 power_cfg_idx = 0;
+	u32 energy_cfg_idx = 0;
+	u32 cmd_mask;
+	int iter;
+
+	/* Psys is available only on CPU0. */
+	if (mgr->client->addr != PECI_BASE_ADDR) {
+		dev_dbg(dev, "not supported CPU index\n");
+		return -ENODEV;
+	}
+
+	cmd_mask = BIT(PECI_CMD_RD_PKG_CFG) | BIT(PECI_CMD_WR_PKG_CFG);
+	if ((mgr->client->adapter->cmd_mask & cmd_mask) != cmd_mask)
+		return -ENODEV;
+
+	/* Psys is supported only on specific CPU generations. */
+	for (iter = 0; iter < ARRAY_SIZE(peci_platformpower_models); ++iter) {
+		if (mgr->gen_info->model == peci_platformpower_models[iter])
+			break;
+	}
+	if (iter == ARRAY_SIZE(peci_platformpower_models)) {
+		dev_dbg(dev, "not supported CPU model\n");
+		return -ENODEV;
+	}
+
+	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	dev_set_drvdata(dev, priv);
+	priv->mgr = mgr;
+	priv->dev = dev;
+	priv->gen_info = mgr->gen_info;
+
+	snprintf(priv->name, PECI_NAME_SIZE, "peci_platformpower.cpu%d",
+		 mgr->client->addr - PECI_BASE_ADDR);
+
+	/* Prepare the right configuration for registration in hwmon. */
+	priv->power_config[power_cfg_idx] = HWMON_P_LABEL |
+		peci_sensor_get_config(peci_platformpower_power_cfg[power_cfg_idx],
+				       ARRAY_SIZE(peci_platformpower_power_cfg[power_cfg_idx]));
+
+	priv->energy_config[energy_cfg_idx] = HWMON_E_LABEL |
+		peci_sensor_get_config(peci_platformpower_energy_cfg[energy_cfg_idx],
+				       ARRAY_SIZE(peci_platformpower_energy_cfg[energy_cfg_idx]));
+
+	priv->info[PECI_PLATFORMPOWER_SENSOR_TYPE_POWER] = &priv->power_info;
+	priv->power_info.type = hwmon_power;
+	priv->power_info.config = priv->power_config;
+
+	priv->info[PECI_PLATFORMPOWER_SENSOR_TYPE_ENERGY] = &priv->energy_info;
+	priv->energy_info.type = hwmon_energy;
+	priv->energy_info.config = priv->energy_config;
+
+	priv->chip.ops = &peci_platformpower_ops;
+	priv->chip.info = priv->info;
+
+	hwmon_dev = devm_hwmon_device_register_with_info(priv->dev, priv->name,
+							 priv, &priv->chip,
+							 NULL);
+
+	if (IS_ERR(hwmon_dev))
+		return PTR_ERR(hwmon_dev);
+
+	dev_dbg(dev, "%s: sensor '%s'\n", dev_name(hwmon_dev),
+		priv->name);
+
+	return 0;
+}
+
+static const struct platform_device_id peci_platformpower_ids[] = {
+	{ .name = "peci-platformpower", .driver_data = 0 },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, peci_platformpower_ids);
+
+static struct platform_driver peci_platformpower_driver = {
+	.probe    = peci_platformpower_probe,
+	.id_table = peci_platformpower_ids,
+	.driver   = { .name = KBUILD_MODNAME, },
+};
+module_platform_driver(peci_platformpower_driver);
+
+MODULE_AUTHOR("Zbigniew Lukwinski <zbigniew.lukwinski@linux.intel.com>");
+MODULE_DESCRIPTION("PECI platformpower driver");
+MODULE_LICENSE("GPL v2");
diff --git a/drivers/mfd/intel-peci-client.c b/drivers/mfd/intel-peci-client.c
index 3eb2c59a2424..34391b698102 100644
--- a/drivers/mfd/intel-peci-client.c
+++ b/drivers/mfd/intel-peci-client.c
@@ -23,6 +23,7 @@ static struct mfd_cell peci_functions[] = {
 	{ .name = "peci-dimmtemp", },
 	{ .name = "peci-cpupower", },
 	{ .name = "peci-dimmpower", },
+	{ .name = "peci-platformpower", },
 };
 
 static const struct cpu_gen_info cpu_gen_info_table[] = {
@@ -62,6 +63,12 @@ static const struct cpu_gen_info cpu_gen_info_table[] = {
 		.core_mask_bits = CORE_MASK_BITS_ON_ICXD,
 		.chan_rank_max  = CHAN_RANK_MAX_ON_ICXD,
 		.dimm_idx_max   = DIMM_IDX_MAX_ON_ICXD },
+	{ /* Sapphire Rapids */
+		.family         = INTEL_FAM6,
+		.model          = INTEL_FAM6_SAPPHIRERAPIDS,
+		.core_mask_bits = CORE_MASK_BITS_ON_SPR,
+		.chan_rank_max  = CHAN_RANK_MAX_ON_SPR,
+		.dimm_idx_max   = DIMM_IDX_MAX_ON_SPR },
 };
 
 static int peci_client_get_cpu_gen_info(struct peci_client_manager *priv)
diff --git a/include/linux/mfd/intel-peci-client.h b/include/linux/mfd/intel-peci-client.h
index 55dec5be3245..e9ad77b19595 100644
--- a/include/linux/mfd/intel-peci-client.h
+++ b/include/linux/mfd/intel-peci-client.h
@@ -20,6 +20,7 @@
 #define INTEL_FAM6_SKYLAKE_XD		0x56
 #define INTEL_FAM6_ICELAKE_X		0x6A
 #define INTEL_FAM6_ICELAKE_XD		0x6C
+#define INTEL_FAM6_SAPPHIRERAPIDS	0x8F
 #endif
 
 #define INTEL_FAM6             6 /* P6 (Pentium Pro and later) */
@@ -48,7 +49,11 @@
 #define CHAN_RANK_MAX_ON_ICXD  4  /* Max number of channel ranks on Icelake D */
 #define DIMM_IDX_MAX_ON_ICXD   2  /* Max DIMM index per channel on Icelake D */
 
-#define CORE_MASK_BITS_MAX     CORE_MASK_BITS_ON_ICX
+#define CORE_MASK_BITS_ON_SPR  64
+#define CHAN_RANK_MAX_ON_SPR   8  /* Max number of channel ranks on Sapphire Rapids */
+#define DIMM_IDX_MAX_ON_SPR    2  /* Max DIMM index per channel on Sapphire Rapids */
+
+#define CORE_MASK_BITS_MAX     CORE_MASK_BITS_ON_SPR
 #define CHAN_RANK_MAX          CHAN_RANK_MAX_ON_HSX
 #define DIMM_IDX_MAX           DIMM_IDX_MAX_ON_HSX
 #define DIMM_NUMS_MAX          (CHAN_RANK_MAX * DIMM_IDX_MAX)
-- 
2.17.1

